### PR TITLE
Peptide level

### PR DESCRIPTION
**Pull request by leeprichman**
* added peptide level input option
* document and test changes added appropriately
* hashed out non-functional line to remove unmatched DAI
* fixed inappropriate edge case check

### CODE REVIEW

File: R/antigen.garnish_predict.R [hunk 1]
```diff
@@ -1785,25 +1788,27 @@ if (assemble & input_type == "transcript"){
...
-        failn <- dt[is.na(mutant_index) | (nchar(pep_mut) - as.numeric(mutant_index) < 1)] %>% nrow
-        warning(paste0("Could not determine mutant index for ", failn, " records."))
-      }
+    # this shouldn't happen, don't need this check, not implemented correctly because
+    # mutant index is a space sep string
+  #     if (dt[is.na(mutant_index) | (nchar(pep_mut) - as.numeric(mutant_index) < 0)] %>%
+  #           nrow > 1){
+  #       failn <- dt[is.na(mutant_index) | (nchar(pep_mut) - as.numeric(mutant_index) < 0)] %>% nrow
+  #       warning(paste0("Could not determine mutant index for ", failn, " records."))
+  #     }
```
[the remark this answers is not in the document]
Comment (leeprichman) [reply R/antigen.garnish_predict.R]: taken care of

File: R/antigen.garnish_predict.R [hunk 2]
```diff
@@ -1840,6 +1845,14 @@ if (generate){
...
 
   # generate a data table of unique variants for peptide generation
+  # this gives index error with frameshift inputs, only occurs when separate_rows is used above
+  # open issue on data.table repo, name exists but column index is invalid:
+  # Error in getindex(x, names(x)[xcols]) :
+  # Internal error: index 'frameshift' exists but is invalid
+  # I randomly fixed by setting key to another column, at some point key was set
+  # to frameshift, this produces error on dtfs and dtnfs creation
+  # data.table::copy at dts creation did not solve this
+    setkey(dts, "sample_id")
```
[the remark this answers is not in the document]
Comment (leeprichman) [reply R/antigen.garnish_predict.R]: I checked the entire function up to this point, no other key is set at any point before this, and nothing that has an expected table order or a key call dt[TRUE] eg exists.   Keys are explicitly set right before they are relied on for table order downstream of this in make_BLAST_uuid and the prediction functions.  This shouldn't be a problem.

File: R/antigen.garnish_predict.R [hunk 3]
```diff
@@ -1810,8 +1809,43 @@ if (assemble & input_type == "peptide"){
 
     dt <- dt[pep_mut %like% "^[ARNDCQEGHILKMFPSTWYV]+$"]
 
+    if ("pep_wt" %chin% names(dt)){
+
+      message("Checking for non-standard AA one-letter codes in \"pep_wt\". Non-standard rows will be discarded.")
+
+      dt <- dt[pep_wt %like% "^[ARNDCQEGHILKMFPSTWYV]+$"]
```
Comment (andrewrech) [inline R/antigen.garnish_predict.R]: @leeprichman why is this repeated inside this loop?

File: R/antigen.garnish_predict.R [hunk 4]
```diff
@@ -1810,8 +1809,43 @@ if (assemble & input_type == "peptide"){
...
+
+      if (nrow(dt) == 0) stop("All rows contained non-standard AAs in either pep_mut or pep_wt columns.")
```
Comment (andrewrech) [inline R/antigen.garnish_predict.R]: I would change this to `any`. In my opinion, if something is wrong about the input, the execution should not proceed, even if it is not just one line, because the output will not be as expected. A warning would be reasonable, but no on reads warnings.

File: R/antigen.garnish_predict.R [hunk 5]
```diff
@@ -1810,8 +1809,43 @@ if (assemble & input_type == "peptide"){
...
+
+      if (any(dt[, mutant_index %>% unique] %like% "\\ ")){
+        message(paste(
```
Comment (andrewrech) [inline R/antigen.garnish_predict.R]: @leeprichman similar to above, I would `stop()` here

File: R/antigen.garnish_predict.R [hunk 6]
```diff
@@ -1810,8 +1809,43 @@ if (assemble & input_type == "peptide"){
...
+          "MNV and frameshifts are not supported in paired mutant wild-type peptide input mode.",
+          "Please provide a single amino acid position as \"mutant_index\" or use \"pep_mut\" input only.",
+          "Offending rows have been discarded.", sep = "\n"))
+
+        dt <- dt[!mutant_index %like% "\\ " & mutant_index != "all"]
+
+        if (nrow(dt) == 0)
+          stop("At least one row must contain a single mutant index for paired mutant and wild-type input.
+          No MNVs or frameshifts in this mode.")
```
Comment (andrewrech) [inline R/antigen.garnish_predict.R]: And then this can be deleted @leeprichman

File: R/antigen.garnish_predict.R [hunk 7]
```diff
@@ -1810,8 +1809,43 @@ if (assemble & input_type == "peptide"){
...
+
+      }
+
+      if (any(dt[, stringr::str_detect(pattern = stringr::fixed(pep_mut), stringr::fixed(pep_wt))])){
+
+        message("Rows where pep_mut contained no mutant sequence have been dropped.")
```
Comment (andrewrech) [inline R/antigen.garnish_predict.R]: Change to warning and print lines? @leeprichman

File: R/antigen.garnish_predict.R (hunk 3 above)
Comment (leeprichman) [reply R/antigen.garnish_predict.R]: same function on different columns, as discussed

File: R/antigen.garnish_predict.R (hunk 4 above)
Comment (leeprichman) [reply R/antigen.garnish_predict.R]: I'm fine with that. I will implement that change.  We will still need the AA whitelist elsewhere in the package for transcript level input because the GRCh/m38 db does have some peptides with non-standard AA afaik.
